iiod/iiod.c

```diff
@@ -106,57 +110,255 @@ static const char *options_descriptions[] = {
+ */
+
+/* Global Data */
+
+static struct avahi_data {
+	AvahiThreadedPoll *poll;
+	AvahiClient *client;
+	AvahiEntryGroup *group;
+	char * name;
+} avahi;
+
+static void create_services(AvahiClient *c);
+static AvahiClient * client_new(void);
+
+static void client_free()
+{
+	/* This also frees the entry group, if any. */
+	if (avahi.client) {
+		avahi_client_free(avahi.client);
+		avahi.client = NULL;
+		avahi.group = NULL;
+	}
+}
+
+static void shutdown_avahi()
+{
+	/* Stop the avahi client, if it's running. */
+	if (avahi.poll)
+		avahi_threaded_poll_stop(avahi.poll);
+
+	/* Clean up the avahi objects. The order is significant. */
+	client_free();
+	if (avahi.poll) {
+		avahi_threaded_poll_free(avahi.poll);
+		avahi.poll = NULL;
+	}
+	if (avahi.name) {
+		INFO("Avahi: Removing service '%s'\n", avahi.name);
+		avahi_free(avahi.name);
+		avahi.name = NULL;
+	}
+}
 
 static void __avahi_group_cb(AvahiEntryGroup *group,
 		AvahiEntryGroupState state, void *d)
 {
+	/* Called whenever the entry group state changes */
+	if (!group)  {
+		ERROR("__avahi_group_cb with no valid group\n");
+		return;
+	}
+
+	avahi.group = group;
+
+	switch (state) {
+		case AVAHI_ENTRY_GROUP_ESTABLISHED :
+			INFO("Avahi: Service '%s' successfully established.\n",
+					avahi.name);
+			break;
+		case AVAHI_ENTRY_GROUP_COLLISION : {
+			char *n;
+			/* A service name collision, pick a new name */
+			n = avahi_alternative_service_name(avahi.name);
+			avahi_free(avahi.name);
+			avahi.name = n;
+			INFO("Avahi: Group Service name collision, renaming service to '%s'\n",
+					avahi.name);
+			create_services(avahi_entry_group_get_client(group));
+			break;
+		}
+		case AVAHI_ENTRY_GROUP_FAILURE :
+			ERROR("Entry group failure: %s\n",
+					avahi_strerror(avahi_client_errno(
+						avahi_entry_group_get_client(group))));
+			break;
+		case AVAHI_ENTRY_GROUP_UNCOMMITED:
+			/* This is normal,
+			 * since we commit things in the create_services()
+			 */
+			DEBUG("Avahi: Group uncommited\n");
+			break;
+		case AVAHI_ENTRY_GROUP_REGISTERING:
+			DEBUG("Avahi: Group registering\n");
+			break;
+	}
 }
 
 static void __avahi_client_cb(AvahiClient *client,
 		AvahiClientState state, void *d)
 {
-	AvahiEntryGroup *group;
-
-	if (state != AVAHI_CLIENT_S_RUNNING)
+	if (!client) {
+		ERROR("__avahi_client_cb with no valid client\n");
 		return;
+	}
 
-	group = avahi_entry_group_new(client, __avahi_group_cb, NULL);
-
-	if (group && !avahi_entry_group_add_service(group,
-			AVAHI_IF_UNSPEC, AVAHI_PROTO_UNSPEC,
-			0, "iio", "_iio._tcp", NULL, NULL, IIOD_PORT, NULL)) {
-		avahi_entry_group_commit(group);
-		INFO("Registered to ZeroConf server %s\n",
-				avahi_client_get_version_string(client));
+	switch (state) {
+		case AVAHI_CLIENT_S_RUNNING:
+			/* Same as AVAHI_SERVER_RUNNING */
+			DEBUG("Avahi: create services\n");
+			/* The server has startup successfully, so create our services */
+			create_services(client);
```

Comment:
is there a chance that `create_services()` requires some cleanup before getting called?
i.e. can the state-machine of avahi do something like `RUNNING, <other-state>, RUNNING`?

Reply:
hmm; this looks the same in the example;
hopefully it's fine

Reply:
There isn't any cleanup, in create_services(), it checks if the group exists, if not it creates it, makes sure the group is empty (the first time being called), and adds the service to the newly created group.

avahi_entry_group_reset(avahi.group); will reset the group the way that the libavahi likes it; which is called in the necessary places (according to the example) before create_services();